src/common/checksumcalculator.h

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (C) 2023 by Oleksandr Zolotov <alex@nextcloud.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+
+#pragma once
+
+#include "ocsynclib.h"
+#include "config.h"
+#include "checksumconsts.h"
+
+#include <QObject>
+#include <QByteArray>
+#include <QFutureWatcher>
+#include <QMutex>
+
+#include <memory>
+
+class QCryptographicHash;
+
+namespace OCC {
+class OCSYNC_EXPORT ChecksumCalculator
+{
+    Q_DISABLE_COPY(ChecksumCalculator)
+
+public:
+    enum class AlgorithmType {
+        Undefined = -1,
+        MD5,
+        SHA1,
+        SHA256,
+        SHA3_256,
+        Adler32,
+    };
+
+    ChecksumCalculator(QSharedPointer<QIODevice> sharedDevice, const QByteArray &checksumTypeName);
+    ~ChecksumCalculator();
+    QByteArray calculate();
```

Comment:
I imagine that you want to add `[[nodiscard]]` as I would not expect this to be useful to someone not reading the returned value